### DISCUSSION TITLE
Audience missing if the client role is applied to same client service account roles

### DISCUSSION THREAD

**Discussion by mxab** [keycloak/keycloak#12971] [Q&A]
Hi,

I just noticed that I have a confidential client X that defines a role A. Client X has also service account enabled and gets its own role A assigned.

If I create a jwt I noticed that client x is missing the client x audience entry. Usually if I would assign any other client the role A the audience is there without any extra mappers.

Is this on purpose so? I know it is maybe not a common use case but why is there an exception for that?

**Comment by mxab** [comment]
Ok it looks like this behavior is very intentional

https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/services/src/main/java/org/keycloak/protocol/oidc/mappers/AudienceResolveProtocolMapper.java#L86

But why? Does this introduce any security issues?

**Reply by mxab** [reply]
according to the git history there was a issue in jira KEYCLOAK-8482 , but I cannot find it

**Comment by eicki** [comment]
I noticed this issue quite a while ago, while I found various notes on this change, the why has not been answered. This is espeically an issue if you use the Spring security Keycloak Adapter, the keycloak.json file only contains a single resource/client which is both used for the redirect as well as for the audience check, so the best practice to use a frontend and backend Client cannot be implemented with this Adapter. I ended up with writing my own audience resolve mapper.

**Comment by sybereal** [comment]
Looking at the [documentation](https://github.com/keycloak/keycloak-documentation/blob/28e907a3faa9061387436e61975cc2ab4b7d0297/server_admin/topics/clients/oidc/con-audience.adoc#automatically-add-audience), this seems to be very much intentional indeed. A workaround and advice is included as well.

![image](https://user-images.githubusercontent.com/103007302/187423517-667a2e2e-ffe2-4da5-b9c5-cf20e7f1c021.png)

**Reply by mxab** [reply]
Thanks for the link to the documentation!

Still it would be interesting why it is not recommended. I guess the idea is that a service is talking to itself is strange, but does this come with any security considerations why they disabled this?
